File: Zig/day02/src/main.zig
const PUZZLE: []const u8 = @embedFile("../../../Inputs/day02.txt");

usingnamespace @import("std");

fn parse(s: []const u8) !ArrayList(usize) {
    var buffer = ArrayList(usize).init(heap.direct_allocator);
    errdefer buffer.deinit();

    var splitter = mem.separate(PUZZLE, ",");

    while (splitter.next()) |w| {
        var num = try fmt.parseInt(usize, w, 10);
        try buffer.append(num);
    }

    return buffer;
}

fn run(program: []usize, noun: usize, verb: usize) usize {
    var ip = usize(0);
    program[1] = noun;
    program[2] = verb;

    while (true) {
        var rhs_place = usize(0);
        var lhs_place = usize(0);
        var result_place = usize(0);

        switch (program[ip]) {
            1...2 => {
                lhs_place = program[ip + 1];
                rhs_place = program[ip + 2];
                result_place = program[ip + 3];
            },
            99 => break,
            else => @panic("unreachable"),
        }

        switch (program[ip]) {
            1 => program[result_place] = program[lhs_place] + program[rhs_place],
            2 => program[result_place] = program[lhs_place] * program[rhs_place],
            else => @panic("unreachable"),
        }

        ip += 4;
    }

    return program[0];
}

fn part1(s: []const u8) !usize {
    var parsed = try parse(s);
    defer parsed.deinit();
    return run(parsed.toSlice(), 12, 2);
}

fn part2(s: []const u8) !usize {
    var magic_number = usize(19690720);

    var parsed = try parse(s);
    defer parsed.deinit();

    var clone = ArrayList(usize).init(heap.direct_allocator);
    defer clone.deinit();

    var noun = usize(0);
    var verb = usize(0);

    while (noun < 99) : ({ noun += 1; }) {
        while (verb < 99) : ({ verb += 1; }) {
            try clone.resize(0);
            try clone.appendSlice(parsed.toSliceConst());

            if (run(clone.toSlice(), noun, verb) == magic_number) {
                return 100 * noun + verb;
            }
        }

        verb = 0;
    }

    @panic("Unreachable");
}

pub fn main() anyerror!void {
    var p1 = try part1(PUZZLE);
    var p2 = try part2(PUZZLE);

    debug.warn("Part 1: {}\nPart 2: {}", p1, p2);
}

File: Rust/day02/src/main.rs
static PUZZLE: &'static str = include_str!(r"..\..\..\Inputs\day02.txt");

fn parse(s: &str) -> Vec<usize> {
    s.split(",")
        .map(|w| w.parse::<usize>().unwrap())
        .collect::<Vec<_>>()
}

#[derive(Eq, PartialEq)]
enum Opcode {
    Add,
    Mul,
    Halt,
}

impl From<usize> for Opcode {
    fn from(n: usize) -> Self {
        match n {
            1 => Self::Add,
            2 => Self::Mul,
            99 => Self::Halt,
            _ => panic!(format!("Invalid opcode: {}", n)),
        }
    }
}

fn run(program: &mut [usize], noun: usize, verb: usize) -> usize {
    let mut ip = 0;
    program[1] = noun;
    program[2] = verb;

    loop {
        let opcode = Opcode::from(program[ip]);
        let (lhs_place, rhs_place, result_place) = match opcode {
            Opcode::Halt => break,
            Opcode::Add | Opcode::Mul => (program[ip + 1], program[ip + 2], program[ip + 3]),
        };

        let result = match opcode {
            Opcode::Add => program[lhs_place] + program[rhs_place],
            Opcode::Mul => program[lhs_place] * program[rhs_place],
            Opcode::Halt => unreachable!(),
        };

        program[result_place] = result;
        ip += 4;
    }

    program[0]
}

fn part1(s: &str) -> usize {
    run(&mut parse(s), 12, 2)
}

fn part2(s: &str) -> usize {
    const MAGIC_NUMBER: usize = 19690720;
    let program = parse(s);

    let mut clone = Vec::with_capacity(program.len());

    (0..=99)
        .flat_map(|noun| (0..=99).map(move |verb| (noun, verb)))
        .find(|(noun, verb)| {
            clone.clear();
            clone.extend_from_slice(&program);

            run(&mut clone, *noun, *verb) == MAGIC_NUMBER
        })
        .map(|(noun, verb)| 100 * noun + verb)
        .unwrap()
}

fn main() {
    let p1 = part1(PUZZLE);
    let p2 = part2(PUZZLE);

    println!("Part 1: {}\nPart 2: {}", p1, p2);
}
